tests/main/snap-quota-thread/task.yaml

```diff
@@ -0,0 +1,117 @@
+  - -ubuntu-core-16-*
+
+prepare: |
+  snap install test-snapd-stressd --edge --devmode
+  snap install test-snapd-curl --edge --devmode
+  snap set system experimental.quota-groups=true
+  tests.cleanup defer snap unset system experimental.quota-groups
+
+restore: |
+  echo "Stop the service"
+  snap stop test-snapd-stressd.spawner
+
+execute: |
+  # Spawner spawns 32 workers per default in the snap, but the total
+  # count of threads expected is 33 including the main thread.
+  SPAWNER_MAX_THREADS=33
+
+  echo "Create a group with the spawner snap in it"
+  snap set-quota group-one --threads=16 test-snapd-stressd
+
+  echo "Starting service"
+  # the reason we start the service afterwards is because the task limit
+  # wont kill any tasks that exceeds the limit, so we test instead that we
+  # can increment the limit. 
+  snap start test-snapd-stressd.spawner
+  sleep 1
+
+  echo "The systemd slice should be active now"
+  sliceName="snap.$(systemd-escape --path group-one).slice"
+  systemctl show --property=ActiveState "$sliceName" | MATCH "ActiveState=active"
+
+  echo "The service should be active"
+  snap services test-snapd-stressd.spawner | MATCH "test-snapd-stressd.spawner\s+disabled\s+active"
+
+  # systemd/kernel have three different locations for the cgroup pids depending
+  # on version
+  echo "The systemd slice should have one process in it now"
+  cgroupsV1OldSystemdProcsFile="/sys/fs/cgroup/pids/$sliceName/snap.test-snapd-stressd.spawner.service/cgroup.procs"
+  cgroupsV1ProcsFile="/sys/fs/cgroup/pids/$sliceName/cgroup.procs"
+  cgroupsV2ProcsFile="/sys/fs/cgroup/$sliceName/snap.test-snapd-stressd.spawner.service/cgroup.procs"
+  if [ -e "$cgroupsV2ProcsFile" ]; then
+      cgroupProcsFile="$cgroupsV2ProcsFile"
+  elif [ -e "$cgroupsV1OldSystemdProcsFile" ]; then
+      cgroupProcsFile="$cgroupsV1OldSystemdProcsFile"
+  elif [ -e "$cgroupsV1ProcsFile" ]; then
+      cgroupProcsFile="$cgroupsV1ProcsFile"
+  else
+      echo "cannot detect cgroup procs file"
+      exit 1
+  fi
+
+  #shellcheck disable=SC2016
+  retry --wait 1 -n 100 --env cgroupProcsFile="$cgroupProcsFile" sh -x -c 'test "$(wc -l < $cgroupProcsFile)" = 1'
+  SERVER_PID=$(cat "$cgroupProcsFile")
+
+  echo "And that process is the main PID for spawner"
+  systemctl show --property=MainPID snap.test-snapd-stressd.spawner.service | MATCH "MainPID=$SERVER_PID"
+
+  echo "And the service is in the Control Group for the slice"
+  # using a regexp for the ControlGroup match here as on older systemd (16.04)
+  # the group is double escaped
+  systemctl show --property=ControlGroup snap.test-snapd-stressd.spawner.service | MATCH 'ControlGroup=/snap.group(.*)one.slice/snap.test-snapd-stressd.spawner.service'
+
+  # snap quota group-one formats the quota usage as a nice human readable 
+  # string, which complicates the comparison here, so instead just grab the 
+  # task usage from the REST API instead
+  echo "Verify the number of threads is correct (=16)"
```

Comment:
`systemd-cgtop` also shows the number of active tasks, if you want to make sure that systemd is living up to our expectations :-)

Reply:
Really good suggestion! I think the current test suffices, because we check the current usage of threads, not the constraint set - and the current usage is the one reported by systemd (by proxy through snap), unless you have very strong feelings about also double checking with systemd-cgtop.

Reply:
Unfortunately, as I wrote above, systemd-cgtop operates on cgroups, and not on systemd units; so, I don't think it can be used in our case. :-(